packages/dependency-extraction-webpack-plugin/lib/index.js

```diff
@@ -104,116 +111,139 @@ class DependencyExtractionWebpackPlugin {
 	apply( compiler ) {
 		this.externalsPlugin.apply( compiler );
 
-		compiler.hooks.emit.tap( this.constructor.name, ( compilation ) => {
-			const {
-				combineAssets,
-				combinedOutputFile,
-				injectPolyfill,
-				outputFormat,
-			} = this.options;
-
-			const combinedAssetsData = {};
-
-			// Process each entry point independently.
-			for ( const [
-				entrypointName,
-				entrypoint,
-			] of compilation.entrypoints.entries() ) {
-				const entrypointExternalizedWpDeps = new Set();
-				if ( injectPolyfill ) {
-					entrypointExternalizedWpDeps.add( 'wp-polyfill' );
+		if ( isWebpack4 ) {
+			compiler.hooks.emit.tap( this.constructor.name, ( compilation ) =>
+				this.addAssets( compilation, compiler )
+			);
+		} else {
+			compiler.hooks.thisCompilation.tap(
+				this.constructor.name,
+				( compilation ) => {
+					compilation.hooks.processAssets.tap(
+						{
+							name: this.constructor.name,
+							stage:
+								webpack.Compilation
+									.PROCESS_ASSETS_STAGE_ADDITIONAL,
+						},
+						() => this.addAssets( compilation, compiler )
+					);
 				}
+			);
+		}
+	}
 
-				const processModule = ( { userRequest } ) => {
-					if ( this.externalizedDeps.has( userRequest ) ) {
-						const scriptDependency = this.mapRequestToDependency(
-							userRequest
-						);
-						entrypointExternalizedWpDeps.add( scriptDependency );
-					}
-				};
-
-				// Search for externalized modules in all chunks.
-				for ( const chunk of entrypoint.chunks ) {
-					for ( const chunkModule of chunk.modulesIterable ) {
-						processModule( chunkModule );
-						// loop through submodules of ConcatenatedModule
-						if ( chunkModule.modules ) {
-							for ( const concatModule of chunkModule.modules ) {
-								processModule( concatModule );
-							}
+	addAssets( compilation, compiler ) {
+		const {
+			combineAssets,
+			combinedOutputFile,
+			injectPolyfill,
+			outputFormat,
+		} = this.options;
+
+		const combinedAssetsData = {};
+
+		// Process each entry point independently.
+		for ( const [
+			entrypointName,
+			entrypoint,
+		] of compilation.entrypoints.entries() ) {
+			const entrypointExternalizedWpDeps = new Set();
+			if ( injectPolyfill ) {
+				entrypointExternalizedWpDeps.add( 'wp-polyfill' );
+			}
+
+			const processModule = ( { userRequest } ) => {
+				if ( this.externalizedDeps.has( userRequest ) ) {
+					const scriptDependency = this.mapRequestToDependency(
+						userRequest
+					);
+					entrypointExternalizedWpDeps.add( scriptDependency );
+				}
+			};
+
+			// Search for externalized modules in all chunks.
+			for ( const chunk of entrypoint.chunks ) {
+				const modulesIterable = isWebpack4
+					? chunk.modulesIterable
+					: compilation.chunkGraph.getChunkModules( chunk );
```

Comment:
There is a difference between webpack 4 and webpack 5 behavior. `chunk.modulesIterable` is sorted by module ID, while `chunkGraph.getChunkModules` returns an unsorted set/array. But it shouldn't matter because `entrypointExternalizedWpDeps` is sorted at the end.